archive/imo/imo1981_q3.lean

```diff
@@ -0,0 +1,235 @@
+Authors: Kevin Lacker
+-/
+
+import data.int.basic
+import data.nat.fib
+import tactic.linarith
+
+/-!
+# IMO 1981 Q3
+
+Determine the maximum value of `m ^ 2 + n ^ 2`, where `m` and `n` are integers in
+`{1, 2, ..., 1981}` and `(n ^ 2 - m * n - m ^ 2) ^ 2 = 1`.
+
+The trick to this problem is that `m` and `n` have to be consecutive Fibonacci numbers,
+because you can reduce any solution to a smaller one using the Fibonacci recurrence.
+-/
+
+/-
+First, define the problem in terms of finding the maximum of a set.
+-/
+
+section
+variable (N : ℕ) -- N = 1981
+
+def in_range (n : ℤ) : Prop := 0 < n ∧ n ≤ N
+
+def problem_predicate (m n : ℤ) : Prop :=
+in_range N m ∧ in_range N n ∧ (n ^ 2 - m * n - m ^ 2) ^ 2 = 1
+
+def specified_set : set ℤ :=
+{k : ℤ | ∃ m : ℤ, ∃ n : ℤ, (k = m ^ 2 + n ^ 2 ∧ problem_predicate N m n)}
+
+/-
+We want to reduce every solution to a smaller solution. Specifically,
+we show that when `(m, n)` is a solution, `(n - m, m)` is also a solution,
+except for the base case of `(1, 1)`.
+-/
+
+lemma pp_m_le_n {m n : ℤ} (h1 : problem_predicate N m n) : m ≤ n :=
+begin
+  obtain (h2|h2) : m ≤ n ∨ ¬ m ≤ n, from em(m ≤ n),
+  { exact h2 },
+  { have h3 : n < m, from not_le.mp h2,
+    unfold problem_predicate at h1,
+    have h4 : n ^ 2 - m * n - m ^ 2 = n * (n - m) - m ^ 2, by ring,
+    rw h4 at h1,
+    have h5 : 0 < n, from h1.right.left.left,
+    have h6 : n * (n - m) - m ^ 2 < -1, by nlinarith,
+    have h7 : (n * (n - m) - m ^ 2) ^ 2 > 1, by nlinarith,
+    exact absurd h1.right.right (ne_of_gt h7) }
+end
+
+lemma pp_eq_imp_1 {n : ℤ} (h1 : problem_predicate N n n) : n = 1 :=
+begin
+  have h2 : (n ^ 2 - n * n - n ^ 2) ^ 2 = 1, from h1.right.right,
+  rw [pow_two, pow_two] at h2,
+  norm_num at h2,
+  have h3 : 0 < n, from h1.left.left,
+  have h4 : 1 ≤ n, by linarith,
+  obtain (h5|h5) : 1 = n ∨ 1 < n, from eq_or_lt_of_le h4,
+  { exact h5.symm },
+  { have h6 : 1 < n * n, by nlinarith,
+    have h7 : 1 < n * n * (n * n), by nlinarith,
+    exact absurd h2.symm (ne_of_lt h7) }
+end
+
+lemma pp_reduction {m n : ℤ} (h1 : problem_predicate N m n) (h2 : 1 < n) :
+problem_predicate N (n - m) m :=
+begin
+  have h3 : (m ^ 2 - (n - m) * m - (n - m) ^ 2) ^ 2 = (n ^ 2 - m * n - m ^ 2) ^ 2, by ring,
+  rw h1.right.right at h3,
+  obtain (h4|h4) : m = n ∨ m < n, from eq_or_lt_of_le (pp_m_le_n N h1),
+  { rw h4 at h1,
+    have h5 : n = 1, from pp_eq_imp_1 N h1,
+    exact absurd h5.symm (ne_of_lt h2) },
+  have h6 : 0 < n - m, from sub_pos.mpr h4,
+  have h7 : n - m < n, from sub_lt_self n h1.left.left,
+  have h8 : n ≤ N, from h1.right.left.right,
+  have h9 : n - m ≤ N, by linarith,
+  exact ⟨⟨h6, h9⟩, ⟨h1.1, h3⟩⟩
+end
+
+/-
+Now we can use induction to show that solutions must be Fibonacci numbers.
+It will be convenient to have the lemmas above in their natural number form.
+-/
+
+def nat_predicate (m n : ℕ) : Prop := problem_predicate N ↑m ↑n
+
+lemma np_m_le_n {m n : ℕ} (h1 : nat_predicate N m n) : m ≤ n :=
```

Comment:
If you want, you can write
```
namespace nat_predicate

lemma m_le_n ...

...

end nat_predicate
```
This will allow you to write `h1.m_le_n` --- so called dot-notation, or projection notation.

I think this leads to shorter and more readable proofs.

Reply:
To be clear: it will have the effect of prefixing all your names with `nat_predicate.`, at which point the "manual" prefix `np_` should be removed.

Reply:
OK